src/box/memtx_tree.cc

```diff
@@ -2475,13 +2547,66 @@ memtx_tree_index_create_read_view(struct index *base)
 	struct space *space = space_by_id(base->def->space_id);
 	assert(space != NULL);
 	memtx_tx_snapshot_cleaner_create(&rv->cleaner, space, base);
+	invalidate_tree_iterator(&rv->dump_iterator);
 	rv->index = index;
 	index_ref(base);
 	memtx_tree_view_create(&rv->tree_view, &index->tree);
 	tree_read_view_reset_key_def(rv);
 	return (struct index_read_view *)rv;
 }
 
+/** Dump the tree data into the sort data file. */
+template<bool USE_HINT>
+static int
+memtx_tree_index_read_view_dump_sort_data(
+	struct index_read_view *base, ssize_t tuple_count,
+	struct memtx_sort_data_writer *sort_data, bool *have_more)
+{
+	assert(!base->def->key_def->for_func_index);
+
+	/* Collect the data to save. */
+	struct tree_read_view<USE_HINT> *rv =
+		(struct tree_read_view<USE_HINT> *)base;
+	struct memtx_tree_data<USE_HINT> *buffer =
+		(struct memtx_tree_data<USE_HINT> *)
+			xmalloc(sizeof(*buffer) * tuple_count);
+	auto _ = make_scoped_guard([buffer] { free(buffer); });
+	ssize_t dumped = 0;
+	while (dumped != tuple_count &&
+	       memtx_tree_view_iterator_next(&rv->tree_view,
+					     &rv->dump_iterator)) {
+		struct memtx_tree_data<USE_HINT> *data =
+			memtx_tree_view_iterator_get_elem(
+				&rv->tree_view, &rv->dump_iterator);
+		if (memtx_tx_snapshot_clarify(&rv->cleaner,
+					      data->tuple) == NULL)
+			continue; /* Only dump visible data. */
+		buffer[dumped++] = *data;
+	}
+
+	/* Write the collected data to the sort data file. */
+	if (memtx_sort_data_writer_put(sort_data, buffer,
+				       sizeof(*buffer), dumped) != 0)
```

Comment:
Do we really need to write `hint` to sortdata? Did we measure performance gain? I suspect that reading twice less data from disk and calculating `hint` for each tuple on recovery would be even faster (hint calculation should be fast because of field map).

If we didn't write hints to sortdata, we wouldn't need these methods - we could simply use `index_read_view_iterator_next_raw` and `index_build_next` methods for dumping and loading data.

P.S. Note that using `field_map` on recovery could complicate multithreaded sortdata load in future - we don't use `field_map` in read-view because of synchronization issues. On the other hand, when loading sortdata, tuple headers and map of tuple formats are not modified at all, so it should be safe to access them.

Reply:
How many tuples and indexes were in your experiment?

Reply:
100M.

Reply:
Let's continue using hints.